sys/include/uri_parser.h

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (C) 2020 HAW Hamburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+/**
+ * @defgroup    sys_uri_parser A minimal, non-destructive URI parser
+ * @ingroup     sys
+ * @brief       A minimal, non-destructive URI parser
+ *
+ * @see https://tools.ietf.org/html/rfc3986
+ *
+ * @{
+ *
+ * @brief       Handler functions for uri_parser
+ * @author      Cenk Gündoğan <cenk.guendogan@haw-hamburg.de>
+ *
+ */
+
+#ifndef URI_PARSER_H
+#define URI_PARSER_H
+
+#include <string.h>
+#include <stdint.h>
+#include <stdbool.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief container that holds all results
+ */
+typedef struct {
+    char *scheme;                 /**< scheme */
+    char *userinfo;               /**< userinfo */
+    char *host;                   /**< host */
+    char *port;                   /**< port */
+    char *path;                   /**< path */
+    char *query;                  /**< query */
+    uint16_t scheme_len;          /**< length of @p scheme */
+    uint16_t userinfo_len;        /**< length of @p userinfo */
+    uint16_t host_len;            /**< length of @p host */
+    uint16_t port_len;            /**< length of @p port */
+    uint16_t path_len;            /**< length of @p path */
+    uint16_t query_len;           /**< length of @p query */
+} uri_parser_result_t;
+
+/**
+ * @brief Checks whether @p uri is absolute or relative
+ *
+ * @param[in]   uri       URI to check
+ * @param[in]   uri_len   Length of @p uri
+ *
+ * @return      true      if @p uri is an absolute URI
+ * @return      false     if @p uri is a relative URI
+ */
+bool uri_parser_is_absolute(const char *uri, size_t uri_len);
+
+/**
+ * @brief Checks whether @p uri is absolute or relative
+ *
+ * @param[in]   uri       NULL-terminated URI to check
+ *
+ * @return      true      if @p uri is an absolute URI
+ * @return      false     if @p uri is a relative URI
+ */
+bool uri_parser_is_absolute_string(const char *uri);
+
+/**
+ * @brief Parse a URI
+ *
+ * @param[out]  result    pointer to a container that will hold the result
+ * @param[in]   uri       URI to parse
+ * @param[in]   uri_len   Length of @p uri
+ *
+ * @return      0         on success
+ * @return      -1        on parsing error
+ */
+int uri_parser_process(uri_parser_result_t *result, const char *uri,
+                       size_t uri_len);
+
+/**
+ * @brief Parse a URI
+ *
+ * @param[out]  result    pointer to a container that will hold the result
+ * @param[in]   uri       NULL-terminated URI to parse
```

Comment:
```suggestion
 * @param[in]   uri       Null-terminated URI to parse
```

or better

```suggestion
 * @param[in]   uri       Zero-terminated URI to parse
```

`NULL` is a constant representing a zero-valued `void` pointer ;-).